### ISSUE TITLE
PHP's password_hash function should be used instead of MySQL "password()"

### ISSUE THREAD

**Issue by b14davbo** [HGustavs/LenaSYS#1964]
When creating a new account, logging in, resetting password etc, we should be using PHP's native password hashing function, see [http://php.net/manual/en/function.password-hash.php](url).

Simple code snippet on how to use

``` php
// get password from post
$password = $_POST["password"];

// hash the password
$user_password = password_hash($password, PASSWORD_DEFAULT);
```

Right now, we're using MySQL password() function which is deprecated as of MySQL 5.7.6.
[http://dev.mysql.com/doc/refman/5.7/en/encryption-functions.html#function_password](url)

**Comment by HGustavs** [comment]
Sounds like a good idea, as this allows us to support more than one database e.g. Postgres.

However, we need a way to back-port passwords since all old passwords will be deprecated also.

Remember that this is a system that is running and that the existing users will be left with non-working passwords when this change is implemented.

Perhaps leaving existing passwords in mysql as is but any new passowrds use php together with code that knows whether that password is old or new??

**Comment by b14davbo** [comment]
@HGustavs That is a good idea. We could have two queries, one for MySQL password() and one for PHP password_hash. If a student with an old password logs in we take the opportunity to hash the password and then log the student in. The other option we have is to force-reset all passwords and send the new passwords via mail.

I also heard that group 1 is trying to get change password functionality working which is nice and a little bit related to this since this should be used there. Setting some labels.

**Comment by b14davbo** [comment]
These files are broken. Needs to be fixed ASAP.
- [https://github.com/HGustavs/LenaSYS/blob/master/DuggaSys/changePassword.php](url)
- [https://github.com/HGustavs/LenaSYS/blob/master/DuggaSys/changePasswordForm.php](url)

The form inside `changePasswordForm.php` has an action that leads to `changepw.php` which doesn't exist in LenaSYS.

**Comment by c14caran** [comment]
There is a simple way of solving this while not breaking anything for existing users.

I have looked in sessions.php and the change needed should be pretty minimal.
As mysql passwords always starts with * and php starts with $ it will be simple to check the entered password against the right hashing type. If the password is a mysql password it should be rehashed as a password_hash and saved in the db, making all user passwords slowly update.

Furthermore, a year after this is implemented the mysql password support could be removed and the database cleared of the old mysql passwords. This will make users that have been inactive for a year unable to login, but it is needed to eventually remove all unsafe passwords.

**Comment by c14caran** [comment]
I have made a try at fixing part of this issue. I want input on the strategy and implementation by you @HGustavs before anything is committed to the codebase seperate from the test branch for this.

Branch for testing: https://github.com/HGustavs/LenaSYS/tree/1964-testlogin
Commit: https://github.com/HGustavs/LenaSYS/commit/87dd44bf9af556204525a29bef857e675a9f4000

By using this commit any person that logs in that still has a mysql password will get their password updated to a php password_hash without any prompting.

**Comment by a15tobli** [comment]
Solved in #3207, closing issue.
